blog/_posts/2017-12-05-ml&pl.md

```diff
@@ -0,0 +1,88 @@
+
+*Programmable semantics* would open up new levels of flexibility, and could be provided by a feature similar to macros. This would allow features like multi-GPU training to be built on top of the core system, by specifying where the code should have pure dataflow semantics (as opposed to imperative semantics, which are more flexible but may include side-effects that are unsafe to optimise). It could also allow the kinds of program manipulation needed by probabilistic programming languages, or the [vectorisation](https://www.cs.cmu.edu/~guyb/papers/Ble90.pdf) (batching) passes usually implemented by hand in NLP models.
+
+As well as the PL community, ML engineers should pay close attention to the traditional Automatic Differentiation (AD) community. ML languages can take inspiration from pioneering work on [languages designed for truly first-class derivative](https://arxiv.org/pdf/1611.03416.pdf) support. Such languages can easily mix symbolic with runtime techniques (helping with the tradeoffs mentioned above), mix forward and reverse mode AD (for improved performance and memory usage), and [differentiate GPU kernels](http://mikeinnes.github.io/2017/08/24/cudanative.html) – all with no loss in performance. Most importantly, the benefits of AD and hardware support should extend to user-written libraries (say, for statistical distributions) rather than [requiring rewrites](http://pytorch.org/docs/master/distributions.html) for compatibility with the framework.
+
+ML research will increasingly need more powerful type systems, user-defined types and more means for extension. Gone are the days when it was enough to hard-code support for strided arrays on NVIDIA GPUs; cutting-edge techniques like [sparse machine learning](https://people.eecs.berkeley.edu/~elghaoui/Pubs/cidu2011_final.pdf), new hardware like [TPUs](https://cloud.google.com/tpu/), [Nervana](https://www.intelnervana.com/) and [FPGAs](https://www.forbes.com/sites/moorinsights/2017/08/28/microsoft-fpga-wins-versus-google-tpus-for-ai/#118733643904), and diverse deployment targets like [ARM chips](http://www.wired.co.uk/article/google-raspberry-pi-ai) or the [iPhone’s CoreML](https://developer.apple.com/documentation/coreml) chip all call for greater levels of flexibility. [Large-scale refactoring of core C++ code](https://github.com/tensorflow/tensorflow/pull/5267/files) for each new development will not scale.
```

Comment:
You might want to mention the growth of mixed precision algorithms https://nickhigham.wordpress.com/2015/10/20/the-rise-of-mixed-precision-arithmetic/

Reply:
Would be great if we can relate it back to ML, but AFAIK most folks are going to less, rather than more, precision there.

Reply:
That's the purpose of mixed precision algorithms. Basically, Float16 is not stable in all parts of the algorithm. So being able to specify that the full algorithm is Float16, but compute the Gamma function using a Float32 and then use that in an activation function as a Float16 is better than requiring the whole algorithm to use Float32 all the way through just to support that one calculation.

Reply:
Isn't that a trivial conversion at the right point?

Reply:
That quick example, yes. But it shouldn't be difficult to come up with a more ML example where you need some vector at a higher precision otherwise you get catastrophic cancellation, but you don't want to subject everything to a higher precision (and optimally you want as much as possible down to Float16). One could even imagine a system that partitions for optimal performance automatically.